### PR TITLE
Improved abs

### PR DESCRIPTION
**Pull request by CodeMaxx**
Fix for #834 
@Sumith1896 @isuruf 
- [ ] abs(y-x) is changed to abs(x-y)
- [x] make abs return evaluated values for all numerical types

### CODE REVIEW

**Comment by isuruf** [comment]
Thanks for working on this.
Does this work like SymPy?
```python
In [1]: from sympy import *

In [2]: var("x y z")
Out[2]: (x, y, z)

In [3]: abs(x-y+1)
Out[3]: Abs(x - y + 1)

In [4]: abs(y-x-1)
Out[4]: Abs(x - y + 1)
```

**Comment by CodeMaxx** [comment]
No not like sympy
For e.g.
`eq(*abs(sub(x,y)),*abs(neg(sub(x,y))));` returns `true`
while
`eq(*abs(sub(x,y)),*abs(sub(y,x)));` returns `false` (How do you suggest we handle this?)

I'm not exactly sure how `could_extract_minus` works but its returning `false` for `y-x` and `true` for `-(x-y)`.

**Comment by isuruf** [comment]
You can look at the code of `could_extract_minus` to see what happens. It returns true for `Mul`s with coefficient negative or negative `Number`s.

Have a look at how SymPy selects `x - y + 1` over `-x + y -1` and implement here.

**Comment by CodeMaxx** [comment]
@isuruf  Added the code for `Complex`. Can review.

File: symengine/functions.cpp [hunk 1]
```diff
@@ -3339,9 +3344,18 @@ RCP<const Basic> abs(const RCP<const Basic> &arg)
         } else {
             return arg_;
         }
+    } else if (is_a<Complex>(*arg)) {
+        RCP<const Complex> arg_ = rcp_static_cast<const Complex>(arg);
+        RCP<const Complex> conjugate = rcp_static_cast<const Complex>(sub((*arg_).real_part(),mul(I,(*arg_).imaginary_part())));
+        return sqrt(abs(((*arg_).mulcomp(*conjugate))));
```
Comment (isuruf) [inline symengine/functions.cpp]: `abs` call here is not needed.
Btw, `sqrt(real*real+imaginary*imaginary)` seems simpler than this.

Comment (CodeMaxx) [inline symengine/functions.cpp]: Writing `sqrt(real*real+imaginary*imaginary)` in terms of `mul()`, `real_part()` etc. also led to a long statement. Also I felt that the return statement looked clearer in the present case.

`abs` call removed. :+1:

**Comment by CodeMaxx** [comment]
@isuruf Sympy is using [signsimp](https://github.com/sympy/sympy/blob/master/sympy/simplify/simplify.py#L322) function being called from [here](https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/complexes.py#L442) ,for `y-x+1` thing.
Its a general function and is being used in lots of places apart from `abs`. Do we have such a function in Symengine? If not how do I implement it?... I tried to understand sympy's code but couldn't understand it fully...will require some help on this.This is the first time I'm trying to port something from sympy.

File: symengine/functions.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/functions.cpp]: I think `sqrt(real*real+imaginary*imaginary)` is faster because you are dealing with `mpq_class`s

**Comment by isuruf** [comment]
It's not there in SymEngine. I don't know the SymPy code as well, but you can look at code and come to the place where the actual logic is. For example `signsimp` calls, [sub_pre](https://github.com/sympy/sympy/blob/master/sympy/simplify/cse_opts.py#L12), which in turn calls [could_extract_minus_sign](https://github.com/sympy/sympy/blob/368a43d0f7a6690d15296f724b572dba01e852a9/sympy/core/expr.py#L2094) which has a handling for `Add` which is not there in SymEngine.

**Comment by CodeMaxx** [comment]
This thing has lots of layers. `could_extract_minus` further calls [sort_key](https://github.com/sympy/sympy/blob/368a43d0f7a6690d15296f724b572dba01e852a9/sympy/core/expr.py#L64) which further calls [default_sort_key](https://github.com/sympy/sympy/blob/master/sympy/core/compatibility.py#L410) . I think its not a good idea to add all this to `Abs` class. We should think of a proper structure for this in symengine and develop similar layered code .

**Comment by isuruf** [comment]
Default sort key in SymEngine is RCPBasicKeyLess. You don't have to change abs, just could_extract_minus.

Let me know if you need to know anything more

File: symengine/tests/basic/test_functions.cpp [hunk 2]
```diff
@@ -2346,6 +2347,9 @@ TEST_CASE("Abs: functions", "[functions]")
     RCP<const Symbol> x = symbol("x");
     RCP<const Symbol> y = symbol("y");
 
+    REQUIRE(eq(*abs(add(i2,mul(I,im1))),*sqrt(integer(5))));
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: Can you fix the formatting here? Space after `,`

**Comment by isuruf** [comment]
Looks good. Can you clean up the history a bit?

**Comment by CodeMaxx** [comment]
@isuruf  I've not yet implemented `abs(y-x)` to `abs(x-y)` . Also I didn't exactly understand what you mean by cleaning up the history.

**Comment by Sumith1896** [comment]
@CodeMaxx  Reduce the number of commits and improve messages

**Comment by CodeMaxx** [comment]
@Sumith1896 Squashing commits is creating problem due to the merges with upstream. Need help with that.

**Comment by isuruf** [comment]
That's fine.
I'll create a new issue for improving `could_extract_minus`

**Comment by isuruf** [comment]
Thanks for the PR

**Comment by Sumith1896** [comment]
With merge commits, it's a bit more work to squash.
This should be fine for now. Thanks @CodeMaxx

**Pull request merged**